sample_metadata/parser/generic_parser.py

```diff
@@ -1267,6 +1284,40 @@ def guess_delimiter_from_filename(filename: str):
 
         raise ValueError(f'Unrecognised extension on file: {filename}')
 
+def group_fastqs_by_common_filename_components(d: dict) -> dict:
+    """Groups key-value pairs by common prefix components and suffix components
+    Where the key is the fastq filename and the value is a tuple of the filename split at R1/R2/1/2
+    
+    Returns a dict with grouping term constructed from extracted prefix+suffix identifiers as key
+    and a list of filenames matching the group as value.
+    Sample input: {
+        'S70_L001_R1_001.fastq.gz': ('S70_L001_', 'R1_001.fastq.gz'),    
+        'S70_L001_R2_001.fastq.gz': ('S70_L001_', 'R2_001.fastq.gz'),
+        'S70_L001_R1.fastq.gz': ('S70_L001_', 'R1.fastq.gz'),    
+        'S70_L001_R2.fastq.gz': ('S70_L001_', 'R1.fastq.gz')
+        }
+    Sample output: {
+        'S70_L001_001.fastq.gz': ['S70_L001_R1_001.fastq.gz', 'S70_L001_R2_001.fastq.gz'],
+        'S70_L001_fastq.gz': ['S70_L001_R1.fastq.gz', 'S70_L001_R2.fastq.gz']
+        }
+    """
+    def key_selector(kv: Tuple[str, str]) -> Tuple[str, str]:
```

Comment:
In Python3.9, you can directly use the type.

```suggestion
    def key_selector(kv: tuple[str, str]) -> tuple[str, str]:
```